object with constructor or destructor cannot be declared 'thread'

Master branch cannot be compiled with VS.

@guolinke @wxchan

```
1>------ Build started: Project: LightGBM, Configuration: DLL x64 ------
1>  application.cpp
1>D:\Users\nekit\Downloads\test\LightGBM\include\LightGBM/network.h(198): error C2483: 'LightGBM::Network::linkers_' : object with constructor or destructor cannot be declared 'thread'
1>D:\Users\nekit\Downloads\test\LightGBM\include\LightGBM/network.h(198): error C2483: 'private: static std::unique_ptr<LightGBM::Linkers,std::default_delete<LightGBM::Linkers> > LightGBM::Network::linkers_' : object with constructor or destructor cannot be declared 'thread'
1>D:\Users\nekit\Downloads\test\LightGBM\include\LightGBM/network.h(200): error C2483: 'LightGBM::Network::bruck_map_' : object with constructor or destructor cannot be declared 'thread'
1>D:\Users\nekit\Downloads\test\LightGBM\include\LightGBM/network.h(200): error C2483: 'private: static LightGBM::BruckMap LightGBM::Network::bruck_map_' : object with constructor or destructor cannot be declared 'thread'
1>D:\Users\nekit\Downloads\test\LightGBM\include\LightGBM/network.h(202): error C2483: 'LightGBM::Network::recursive_halving_map_' : object with constructor or destructor cannot be declared 'thread'
1>D:\Users\nekit\Downloads\test\LightGBM\include\LightGBM/network.h(202): error C2483: 'private: static LightGBM::RecursiveHalvingMap LightGBM::Network::recursive_halving_map_' : object with constructor or destructor cannot be declared 'thread'
1>D:\Users\nekit\Downloads\test\LightGBM\include\LightGBM/network.h(204): error C2483: 'LightGBM::Network::block_start_' : object with constructor or destructor cannot be declared 'thread'
1>D:\Users\nekit\Downloads\test\LightGBM\include\LightGBM/network.h(204): error C2483: 'private: static std::vector<int,std::allocator<int> > LightGBM::Network::block_start_' : object with constructor or destructor cannot be declared 'thread'
1>D:\Users\nekit\Downloads\test\LightGBM\include\LightGBM/network.h(206): error C2483: 'LightGBM::Network::block_len_' : object with constructor or destructor cannot be declared 'thread'
1>D:\Users\nekit\Downloads\test\LightGBM\include\LightGBM/network.h(206): error C2483: 'private: static std::vector<int,std::allocator<int> > LightGBM::Network::block_len_' : object with constructor or destructor cannot be declared 'thread'
1>D:\Users\nekit\Downloads\test\LightGBM\include\LightGBM/network.h(208): error C2483: 'LightGBM::Network::buffer_' : object with constructor or destructor cannot be declared 'thread'
1>D:\Users\nekit\Downloads\test\LightGBM\include\LightGBM/network.h(208): error C2483: 'private: static std::vector<char,std::allocator<char> > LightGBM::Network::buffer_' : object with constructor or destructor cannot be declared 'thread'
1>  boosting.cpp
1>  gbdt.cpp
1>D:\Users\nekit\Downloads\test\LightGBM\include\LightGBM/network.h(198): error C2483: 'LightGBM::Network::linkers_' : object with constructor or destructor cannot be declared 'thread'
1>D:\Users\nekit\Downloads\test\LightGBM\include\LightGBM/network.h(198): error C2483: 'private: static std::unique_ptr<LightGBM::Linkers,std::default_delete<LightGBM::Linkers> > LightGBM::Network::linkers_' : object with constructor or destructor cannot be declared 'thread'
1>D:\Users\nekit\Downloads\test\LightGBM\include\LightGBM/network.h(200): error C2483: 'LightGBM::Network::bruck_map_' : object with constructor or destructor cannot be declared 'thread'
1>D:\Users\nekit\Downloads\test\LightGBM\include\LightGBM/network.h(200): error C2483: 'private: static LightGBM::BruckMap LightGBM::Network::bruck_map_' : object with constructor or destructor cannot be declared 'thread'
1>D:\Users\nekit\Downloads\test\LightGBM\include\LightGBM/network.h(202): error C2483: 'LightGBM::Network::recursive_halving_map_' : object with constructor or destructor cannot be declared 'thread'
1>D:\Users\nekit\Downloads\test\LightGBM\include\LightGBM/network.h(202): error C2483: 'private: static LightGBM::RecursiveHalvingMap LightGBM::Network::recursive_halving_map_' : object with constructor or destructor cannot be declared 'thread'
1>D:\Users\nekit\Downloads\test\LightGBM\include\LightGBM/network.h(204): error C2483: 'LightGBM::Network::block_start_' : object with constructor or destructor cannot be declared 'thread'
1>D:\Users\nekit\Downloads\test\LightGBM\include\LightGBM/network.h(204): error C2483: 'private: static std::vector<int,std::allocator<int> > LightGBM::Network::block_start_' : object with constructor or destructor cannot be declared 'thread'
1>D:\Users\nekit\Downloads\test\LightGBM\include\LightGBM/network.h(206): error C2483: 'LightGBM::Network::block_len_' : object with constructor or destructor cannot be declared 'thread'
1>D:\Users\nekit\Downloads\test\LightGBM\include\LightGBM/network.h(206): error C2483: 'private: static std::vector<int,std::allocator<int> > LightGBM::Network::block_len_' : object with constructor or destructor cannot be declared 'thread'
1>D:\Users\nekit\Downloads\test\LightGBM\include\LightGBM/network.h(208): error C2483: 'LightGBM::Network::buffer_' : object with constructor or destructor cannot be declared 'thread'
1>D:\Users\nekit\Downloads\test\LightGBM\include\LightGBM/network.h(208): error C2483: 'private: static std::vector<char,std::allocator<char> > LightGBM::Network::buffer_' : object with constructor or destructor cannot be declared 'thread'
1>  gbdt_model_text.cpp
1>  gbdt_prediction.cpp
1>  prediction_early_stop.cpp
1>  c_api.cpp
1>D:\Users\nekit\Downloads\test\LightGBM\include\LightGBM/network.h(198): error C2483: 'LightGBM::Network::linkers_' : object with constructor or destructor cannot be declared 'thread'
1>D:\Users\nekit\Downloads\test\LightGBM\include\LightGBM/network.h(198): error C2483: 'private: static std::unique_ptr<LightGBM::Linkers,std::default_delete<LightGBM::Linkers> > LightGBM::Network::linkers_' : object with constructor or destructor cannot be declared 'thread'
1>D:\Users\nekit\Downloads\test\LightGBM\include\LightGBM/network.h(200): error C2483: 'LightGBM::Network::bruck_map_' : object with constructor or destructor cannot be declared 'thread'
1>D:\Users\nekit\Downloads\test\LightGBM\include\LightGBM/network.h(200): error C2483: 'private: static LightGBM::BruckMap LightGBM::Network::bruck_map_' : object with constructor or destructor cannot be declared 'thread'
1>D:\Users\nekit\Downloads\test\LightGBM\include\LightGBM/network.h(202): error C2483: 'LightGBM::Network::recursive_halving_map_' : object with constructor or destructor cannot be declared 'thread'
1>D:\Users\nekit\Downloads\test\LightGBM\include\LightGBM/network.h(202): error C2483: 'private: static LightGBM::RecursiveHalvingMap LightGBM::Network::recursive_halving_map_' : object with constructor or destructor cannot be declared 'thread'
1>D:\Users\nekit\Downloads\test\LightGBM\include\LightGBM/network.h(204): error C2483: 'LightGBM::Network::block_start_' : object with constructor or destructor cannot be declared 'thread'
1>D:\Users\nekit\Downloads\test\LightGBM\include\LightGBM/network.h(204): error C2483: 'private: static std::vector<int,std::allocator<int> > LightGBM::Network::block_start_' : object with constructor or destructor cannot be declared 'thread'
1>D:\Users\nekit\Downloads\test\LightGBM\include\LightGBM/network.h(206): error C2483: 'LightGBM::Network::block_len_' : object with constructor or destructor cannot be declared 'thread'
1>D:\Users\nekit\Downloads\test\LightGBM\include\LightGBM/network.h(206): error C2483: 'private: static std::vector<int,std::allocator<int> > LightGBM::Network::block_len_' : object with constructor or destructor cannot be declared 'thread'
1>D:\Users\nekit\Downloads\test\LightGBM\include\LightGBM/network.h(208): error C2483: 'LightGBM::Network::buffer_' : object with constructor or destructor cannot be declared 'thread'
1>D:\Users\nekit\Downloads\test\LightGBM\include\LightGBM/network.h(208): error C2483: 'private: static std::vector<char,std::allocator<char> > LightGBM::Network::buffer_' : object with constructor or destructor cannot be declared 'thread'
1>  bin.cpp
1>  config.cpp
1>  dataset.cpp
1>  dataset_loader.cpp
1>D:\Users\nekit\Downloads\test\LightGBM\include\LightGBM/network.h(198): error C2483: 'LightGBM::Network::linkers_' : object with constructor or destructor cannot be declared 'thread'
1>D:\Users\nekit\Downloads\test\LightGBM\include\LightGBM/network.h(198): error C2483: 'private: static std::unique_ptr<LightGBM::Linkers,std::default_delete<LightGBM::Linkers> > LightGBM::Network::linkers_' : object with constructor or destructor cannot be declared 'thread'
1>D:\Users\nekit\Downloads\test\LightGBM\include\LightGBM/network.h(200): error C2483: 'LightGBM::Network::bruck_map_' : object with constructor or destructor cannot be declared 'thread'
1>D:\Users\nekit\Downloads\test\LightGBM\include\LightGBM/network.h(200): error C2483: 'private: static LightGBM::BruckMap LightGBM::Network::bruck_map_' : object with constructor or destructor cannot be declared 'thread'
1>D:\Users\nekit\Downloads\test\LightGBM\include\LightGBM/network.h(202): error C2483: 'LightGBM::Network::recursive_halving_map_' : object with constructor or destructor cannot be declared 'thread'
1>D:\Users\nekit\Downloads\test\LightGBM\include\LightGBM/network.h(202): error C2483: 'private: static LightGBM::RecursiveHalvingMap LightGBM::Network::recursive_halving_map_' : object with constructor or destructor cannot be declared 'thread'
1>D:\Users\nekit\Downloads\test\LightGBM\include\LightGBM/network.h(204): error C2483: 'LightGBM::Network::block_start_' : object with constructor or destructor cannot be declared 'thread'
1>D:\Users\nekit\Downloads\test\LightGBM\include\LightGBM/network.h(204): error C2483: 'private: static std::vector<int,std::allocator<int> > LightGBM::Network::block_start_' : object with constructor or destructor cannot be declared 'thread'
1>D:\Users\nekit\Downloads\test\LightGBM\include\LightGBM/network.h(206): error C2483: 'LightGBM::Network::block_len_' : object with constructor or destructor cannot be declared 'thread'
1>D:\Users\nekit\Downloads\test\LightGBM\include\LightGBM/network.h(206): error C2483: 'private: static std::vector<int,std::allocator<int> > LightGBM::Network::block_len_' : object with constructor or destructor cannot be declared 'thread'
1>D:\Users\nekit\Downloads\test\LightGBM\include\LightGBM/network.h(208): error C2483: 'LightGBM::Network::buffer_' : object with constructor or destructor cannot be declared 'thread'
1>D:\Users\nekit\Downloads\test\LightGBM\include\LightGBM/network.h(208): error C2483: 'private: static std::vector<char,std::allocator<char> > LightGBM::Network::buffer_' : object with constructor or destructor cannot be declared 'thread'
1>  metadata.cpp
1>  parser.cpp
1>  tree.cpp
1>  lightgbm_R.cpp
1>  dcg_calculator.cpp
1>  metric.cpp
1>  network.cpp
1>D:\Users\nekit\Downloads\test\LightGBM\include\LightGBM/network.h(198): error C2483: 'LightGBM::Network::linkers_' : object with constructor or destructor cannot be declared 'thread'
1>D:\Users\nekit\Downloads\test\LightGBM\include\LightGBM/network.h(198): error C2483: 'private: static std::unique_ptr<LightGBM::Linkers,std::default_delete<LightGBM::Linkers> > LightGBM::Network::linkers_' : object with constructor or destructor cannot be declared 'thread'
1>D:\Users\nekit\Downloads\test\LightGBM\include\LightGBM/network.h(200): error C2483: 'LightGBM::Network::bruck_map_' : object with constructor or destructor cannot be declared 'thread'
1>D:\Users\nekit\Downloads\test\LightGBM\include\LightGBM/network.h(200): error C2483: 'private: static LightGBM::BruckMap LightGBM::Network::bruck_map_' : object with constructor or destructor cannot be declared 'thread'
1>D:\Users\nekit\Downloads\test\LightGBM\include\LightGBM/network.h(202): error C2483: 'LightGBM::Network::recursive_halving_map_' : object with constructor or destructor cannot be declared 'thread'
1>D:\Users\nekit\Downloads\test\LightGBM\include\LightGBM/network.h(202): error C2483: 'private: static LightGBM::RecursiveHalvingMap LightGBM::Network::recursive_halving_map_' : object with constructor or destructor cannot be declared 'thread'
1>D:\Users\nekit\Downloads\test\LightGBM\include\LightGBM/network.h(204): error C2483: 'LightGBM::Network::block_start_' : object with constructor or destructor cannot be declared 'thread'
1>D:\Users\nekit\Downloads\test\LightGBM\include\LightGBM/network.h(204): error C2483: 'private: static std::vector<int,std::allocator<int> > LightGBM::Network::block_start_' : object with constructor or destructor cannot be declared 'thread'
1>D:\Users\nekit\Downloads\test\LightGBM\include\LightGBM/network.h(206): error C2483: 'LightGBM::Network::block_len_' : object with constructor or destructor cannot be declared 'thread'
1>D:\Users\nekit\Downloads\test\LightGBM\include\LightGBM/network.h(206): error C2483: 'private: static std::vector<int,std::allocator<int> > LightGBM::Network::block_len_' : object with constructor or destructor cannot be declared 'thread'
1>D:\Users\nekit\Downloads\test\LightGBM\include\LightGBM/network.h(208): error C2483: 'LightGBM::Network::buffer_' : object with constructor or destructor cannot be declared 'thread'
1>D:\Users\nekit\Downloads\test\LightGBM\include\LightGBM/network.h(208): error C2483: 'private: static std::vector<char,std::allocator<char> > LightGBM::Network::buffer_' : object with constructor or destructor cannot be declared 'thread'
1>..\src\network\network.cpp(15): error C2483: 'LightGBM::Network::linkers_' : object with constructor or destructor cannot be declared 'thread'
1>..\src\network\network.cpp(16): error C2483: 'LightGBM::Network::bruck_map_' : object with constructor or destructor cannot be declared 'thread'
1>..\src\network\network.cpp(17): error C2483: 'LightGBM::Network::recursive_halving_map_' : object with constructor or destructor cannot be declared 'thread'
1>..\src\network\network.cpp(18): error C2483: 'LightGBM::Network::block_start_' : object with constructor or destructor cannot be declared 'thread'
1>..\src\network\network.cpp(19): error C2483: 'LightGBM::Network::block_len_' : object with constructor or destructor cannot be declared 'thread'
1>..\src\network\network.cpp(21): error C2483: 'LightGBM::Network::buffer_' : object with constructor or destructor cannot be declared 'thread'
1>  linkers_mpi.cpp
1>  linkers_socket.cpp
1>D:\Users\nekit\Downloads\test\LightGBM\include\LightGBM/network.h(198): error C2483: 'LightGBM::Network::linkers_' : object with constructor or destructor cannot be declared 'thread'
1>D:\Users\nekit\Downloads\test\LightGBM\include\LightGBM/network.h(198): error C2483: 'private: static std::unique_ptr<LightGBM::Linkers,std::default_delete<LightGBM::Linkers> > LightGBM::Network::linkers_' : object with constructor or destructor cannot be declared 'thread'
1>D:\Users\nekit\Downloads\test\LightGBM\include\LightGBM/network.h(200): error C2483: 'LightGBM::Network::bruck_map_' : object with constructor or destructor cannot be declared 'thread'
1>D:\Users\nekit\Downloads\test\LightGBM\include\LightGBM/network.h(200): error C2483: 'private: static LightGBM::BruckMap LightGBM::Network::bruck_map_' : object with constructor or destructor cannot be declared 'thread'
1>D:\Users\nekit\Downloads\test\LightGBM\include\LightGBM/network.h(202): error C2483: 'LightGBM::Network::recursive_halving_map_' : object with constructor or destructor cannot be declared 'thread'
1>D:\Users\nekit\Downloads\test\LightGBM\include\LightGBM/network.h(202): error C2483: 'private: static LightGBM::RecursiveHalvingMap LightGBM::Network::recursive_halving_map_' : object with constructor or destructor cannot be declared 'thread'
1>D:\Users\nekit\Downloads\test\LightGBM\include\LightGBM/network.h(204): error C2483: 'LightGBM::Network::block_start_' : object with constructor or destructor cannot be declared 'thread'
1>D:\Users\nekit\Downloads\test\LightGBM\include\LightGBM/network.h(204): error C2483: 'private: static std::vector<int,std::allocator<int> > LightGBM::Network::block_start_' : object with constructor or destructor cannot be declared 'thread'
1>D:\Users\nekit\Downloads\test\LightGBM\include\LightGBM/network.h(206): error C2483: 'LightGBM::Network::block_len_' : object with constructor or destructor cannot be declared 'thread'
1>D:\Users\nekit\Downloads\test\LightGBM\include\LightGBM/network.h(206): error C2483: 'private: static std::vector<int,std::allocator<int> > LightGBM::Network::block_len_' : object with constructor or destructor cannot be declared 'thread'
1>D:\Users\nekit\Downloads\test\LightGBM\include\LightGBM/network.h(208): error C2483: 'LightGBM::Network::buffer_' : object with constructor or destructor cannot be declared 'thread'
1>D:\Users\nekit\Downloads\test\LightGBM\include\LightGBM/network.h(208): error C2483: 'private: static std::vector<char,std::allocator<char> > LightGBM::Network::buffer_' : object with constructor or destructor cannot be declared 'thread'
1>  Compiling...
1>  linker_topo.cpp
1>D:\Users\nekit\Downloads\test\LightGBM\include\LightGBM/network.h(198): error C2483: 'LightGBM::Network::linkers_' : object with constructor or destructor cannot be declared 'thread'
1>D:\Users\nekit\Downloads\test\LightGBM\include\LightGBM/network.h(198): error C2483: 'private: static std::unique_ptr<LightGBM::Linkers,std::default_delete<LightGBM::Linkers> > LightGBM::Network::linkers_' : object with constructor or destructor cannot be declared 'thread'
1>D:\Users\nekit\Downloads\test\LightGBM\include\LightGBM/network.h(200): error C2483: 'LightGBM::Network::bruck_map_' : object with constructor or destructor cannot be declared 'thread'
1>D:\Users\nekit\Downloads\test\LightGBM\include\LightGBM/network.h(200): error C2483: 'private: static LightGBM::BruckMap LightGBM::Network::bruck_map_' : object with constructor or destructor cannot be declared 'thread'
1>D:\Users\nekit\Downloads\test\LightGBM\include\LightGBM/network.h(202): error C2483: 'LightGBM::Network::recursive_halving_map_' : object with constructor or destructor cannot be declared 'thread'
1>D:\Users\nekit\Downloads\test\LightGBM\include\LightGBM/network.h(202): error C2483: 'private: static LightGBM::RecursiveHalvingMap LightGBM::Network::recursive_halving_map_' : object with constructor or destructor cannot be declared 'thread'
1>D:\Users\nekit\Downloads\test\LightGBM\include\LightGBM/network.h(204): error C2483: 'LightGBM::Network::block_start_' : object with constructor or destructor cannot be declared 'thread'
1>D:\Users\nekit\Downloads\test\LightGBM\include\LightGBM/network.h(204): error C2483: 'private: static std::vector<int,std::allocator<int> > LightGBM::Network::block_start_' : object with constructor or destructor cannot be declared 'thread'
1>D:\Users\nekit\Downloads\test\LightGBM\include\LightGBM/network.h(206): error C2483: 'LightGBM::Network::block_len_' : object with constructor or destructor cannot be declared 'thread'
1>D:\Users\nekit\Downloads\test\LightGBM\include\LightGBM/network.h(206): error C2483: 'private: static std::vector<int,std::allocator<int> > LightGBM::Network::block_len_' : object with constructor or destructor cannot be declared 'thread'
1>D:\Users\nekit\Downloads\test\LightGBM\include\LightGBM/network.h(208): error C2483: 'LightGBM::Network::buffer_' : object with constructor or destructor cannot be declared 'thread'
1>D:\Users\nekit\Downloads\test\LightGBM\include\LightGBM/network.h(208): error C2483: 'private: static std::vector<char,std::allocator<char> > LightGBM::Network::buffer_' : object with constructor or destructor cannot be declared 'thread'
1>  objective_function.cpp
1>  main.cpp
1>  data_parallel_tree_learner.cpp
1>D:\Users\nekit\Downloads\test\LightGBM\include\LightGBM/network.h(198): error C2483: 'LightGBM::Network::linkers_' : object with constructor or destructor cannot be declared 'thread'
1>D:\Users\nekit\Downloads\test\LightGBM\include\LightGBM/network.h(198): error C2483: 'private: static std::unique_ptr<LightGBM::Linkers,std::default_delete<LightGBM::Linkers> > LightGBM::Network::linkers_' : object with constructor or destructor cannot be declared 'thread'
1>D:\Users\nekit\Downloads\test\LightGBM\include\LightGBM/network.h(200): error C2483: 'LightGBM::Network::bruck_map_' : object with constructor or destructor cannot be declared 'thread'
1>D:\Users\nekit\Downloads\test\LightGBM\include\LightGBM/network.h(200): error C2483: 'private: static LightGBM::BruckMap LightGBM::Network::bruck_map_' : object with constructor or destructor cannot be declared 'thread'
1>D:\Users\nekit\Downloads\test\LightGBM\include\LightGBM/network.h(202): error C2483: 'LightGBM::Network::recursive_halving_map_' : object with constructor or destructor cannot be declared 'thread'
1>D:\Users\nekit\Downloads\test\LightGBM\include\LightGBM/network.h(202): error C2483: 'private: static LightGBM::RecursiveHalvingMap LightGBM::Network::recursive_halving_map_' : object with constructor or destructor cannot be declared 'thread'
1>D:\Users\nekit\Downloads\test\LightGBM\include\LightGBM/network.h(204): error C2483: 'LightGBM::Network::block_start_' : object with constructor or destructor cannot be declared 'thread'
1>D:\Users\nekit\Downloads\test\LightGBM\include\LightGBM/network.h(204): error C2483: 'private: static std::vector<int,std::allocator<int> > LightGBM::Network::block_start_' : object with constructor or destructor cannot be declared 'thread'
1>D:\Users\nekit\Downloads\test\LightGBM\include\LightGBM/network.h(206): error C2483: 'LightGBM::Network::block_len_' : object with constructor or destructor cannot be declared 'thread'
1>D:\Users\nekit\Downloads\test\LightGBM\include\LightGBM/network.h(206): error C2483: 'private: static std::vector<int,std::allocator<int> > LightGBM::Network::block_len_' : object with constructor or destructor cannot be declared 'thread'
1>D:\Users\nekit\Downloads\test\LightGBM\include\LightGBM/network.h(208): error C2483: 'LightGBM::Network::buffer_' : object with constructor or destructor cannot be declared 'thread'
1>D:\Users\nekit\Downloads\test\LightGBM\include\LightGBM/network.h(208): error C2483: 'private: static std::vector<char,std::allocator<char> > LightGBM::Network::buffer_' : object with constructor or destructor cannot be declared 'thread'
1>  feature_parallel_tree_learner.cpp
1>D:\Users\nekit\Downloads\test\LightGBM\include\LightGBM/network.h(198): error C2483: 'LightGBM::Network::linkers_' : object with constructor or destructor cannot be declared 'thread'
1>D:\Users\nekit\Downloads\test\LightGBM\include\LightGBM/network.h(198): error C2483: 'private: static std::unique_ptr<LightGBM::Linkers,std::default_delete<LightGBM::Linkers> > LightGBM::Network::linkers_' : object with constructor or destructor cannot be declared 'thread'
1>D:\Users\nekit\Downloads\test\LightGBM\include\LightGBM/network.h(200): error C2483: 'LightGBM::Network::bruck_map_' : object with constructor or destructor cannot be declared 'thread'
1>D:\Users\nekit\Downloads\test\LightGBM\include\LightGBM/network.h(200): error C2483: 'private: static LightGBM::BruckMap LightGBM::Network::bruck_map_' : object with constructor or destructor cannot be declared 'thread'
1>D:\Users\nekit\Downloads\test\LightGBM\include\LightGBM/network.h(202): error C2483: 'LightGBM::Network::recursive_halving_map_' : object with constructor or destructor cannot be declared 'thread'
1>D:\Users\nekit\Downloads\test\LightGBM\include\LightGBM/network.h(202): error C2483: 'private: static LightGBM::RecursiveHalvingMap LightGBM::Network::recursive_halving_map_' : object with constructor or destructor cannot be declared 'thread'
1>D:\Users\nekit\Downloads\test\LightGBM\include\LightGBM/network.h(204): error C2483: 'LightGBM::Network::block_start_' : object with constructor or destructor cannot be declared 'thread'
1>D:\Users\nekit\Downloads\test\LightGBM\include\LightGBM/network.h(204): error C2483: 'private: static std::vector<int,std::allocator<int> > LightGBM::Network::block_start_' : object with constructor or destructor cannot be declared 'thread'
1>D:\Users\nekit\Downloads\test\LightGBM\include\LightGBM/network.h(206): error C2483: 'LightGBM::Network::block_len_' : object with constructor or destructor cannot be declared 'thread'
1>D:\Users\nekit\Downloads\test\LightGBM\include\LightGBM/network.h(206): error C2483: 'private: static std::vector<int,std::allocator<int> > LightGBM::Network::block_len_' : object with constructor or destructor cannot be declared 'thread'
1>D:\Users\nekit\Downloads\test\LightGBM\include\LightGBM/network.h(208): error C2483: 'LightGBM::Network::buffer_' : object with constructor or destructor cannot be declared 'thread'
1>D:\Users\nekit\Downloads\test\LightGBM\include\LightGBM/network.h(208): error C2483: 'private: static std::vector<char,std::allocator<char> > LightGBM::Network::buffer_' : object with constructor or destructor cannot be declared 'thread'
1>  serial_tree_learner.cpp
1>  tree_learner.cpp
1>D:\Users\nekit\Downloads\test\LightGBM\include\LightGBM/network.h(198): error C2483: 'LightGBM::Network::linkers_' : object with constructor or destructor cannot be declared 'thread'
1>D:\Users\nekit\Downloads\test\LightGBM\include\LightGBM/network.h(198): error C2483: 'private: static std::unique_ptr<LightGBM::Linkers,std::default_delete<LightGBM::Linkers> > LightGBM::Network::linkers_' : object with constructor or destructor cannot be declared 'thread'
1>D:\Users\nekit\Downloads\test\LightGBM\include\LightGBM/network.h(200): error C2483: 'LightGBM::Network::bruck_map_' : object with constructor or destructor cannot be declared 'thread'
1>D:\Users\nekit\Downloads\test\LightGBM\include\LightGBM/network.h(200): error C2483: 'private: static LightGBM::BruckMap LightGBM::Network::bruck_map_' : object with constructor or destructor cannot be declared 'thread'
1>D:\Users\nekit\Downloads\test\LightGBM\include\LightGBM/network.h(202): error C2483: 'LightGBM::Network::recursive_halving_map_' : object with constructor or destructor cannot be declared 'thread'
1>D:\Users\nekit\Downloads\test\LightGBM\include\LightGBM/network.h(202): error C2483: 'private: static LightGBM::RecursiveHalvingMap LightGBM::Network::recursive_halving_map_' : object with constructor or destructor cannot be declared 'thread'
1>D:\Users\nekit\Downloads\test\LightGBM\include\LightGBM/network.h(204): error C2483: 'LightGBM::Network::block_start_' : object with constructor or destructor cannot be declared 'thread'
1>D:\Users\nekit\Downloads\test\LightGBM\include\LightGBM/network.h(204): error C2483: 'private: static std::vector<int,std::allocator<int> > LightGBM::Network::block_start_' : object with constructor or destructor cannot be declared 'thread'
1>D:\Users\nekit\Downloads\test\LightGBM\include\LightGBM/network.h(206): error C2483: 'LightGBM::Network::block_len_' : object with constructor or destructor cannot be declared 'thread'
1>D:\Users\nekit\Downloads\test\LightGBM\include\LightGBM/network.h(206): error C2483: 'private: static std::vector<int,std::allocator<int> > LightGBM::Network::block_len_' : object with constructor or destructor cannot be declared 'thread'
1>D:\Users\nekit\Downloads\test\LightGBM\include\LightGBM/network.h(208): error C2483: 'LightGBM::Network::buffer_' : object with constructor or destructor cannot be declared 'thread'
1>D:\Users\nekit\Downloads\test\LightGBM\include\LightGBM/network.h(208): error C2483: 'private: static std::vector<char,std::allocator<char> > LightGBM::Network::buffer_' : object with constructor or destructor cannot be declared 'thread'
1>  voting_parallel_tree_learner.cpp
1>D:\Users\nekit\Downloads\test\LightGBM\include\LightGBM/network.h(198): error C2483: 'LightGBM::Network::linkers_' : object with constructor or destructor cannot be declared 'thread'
1>D:\Users\nekit\Downloads\test\LightGBM\include\LightGBM/network.h(198): error C2483: 'private: static std::unique_ptr<LightGBM::Linkers,std::default_delete<LightGBM::Linkers> > LightGBM::Network::linkers_' : object with constructor or destructor cannot be declared 'thread'
1>D:\Users\nekit\Downloads\test\LightGBM\include\LightGBM/network.h(200): error C2483: 'LightGBM::Network::bruck_map_' : object with constructor or destructor cannot be declared 'thread'
1>D:\Users\nekit\Downloads\test\LightGBM\include\LightGBM/network.h(200): error C2483: 'private: static LightGBM::BruckMap LightGBM::Network::bruck_map_' : object with constructor or destructor cannot be declared 'thread'
1>D:\Users\nekit\Downloads\test\LightGBM\include\LightGBM/network.h(202): error C2483: 'LightGBM::Network::recursive_halving_map_' : object with constructor or destructor cannot be declared 'thread'
1>D:\Users\nekit\Downloads\test\LightGBM\include\LightGBM/network.h(202): error C2483: 'private: static LightGBM::RecursiveHalvingMap LightGBM::Network::recursive_halving_map_' : object with constructor or destructor cannot be declared 'thread'
1>D:\Users\nekit\Downloads\test\LightGBM\include\LightGBM/network.h(204): error C2483: 'LightGBM::Network::block_start_' : object with constructor or destructor cannot be declared 'thread'
1>D:\Users\nekit\Downloads\test\LightGBM\include\LightGBM/network.h(204): error C2483: 'private: static std::vector<int,std::allocator<int> > LightGBM::Network::block_start_' : object with constructor or destructor cannot be declared 'thread'
1>D:\Users\nekit\Downloads\test\LightGBM\include\LightGBM/network.h(206): error C2483: 'LightGBM::Network::block_len_' : object with constructor or destructor cannot be declared 'thread'
1>D:\Users\nekit\Downloads\test\LightGBM\include\LightGBM/network.h(206): error C2483: 'private: static std::vector<int,std::allocator<int> > LightGBM::Network::block_len_' : object with constructor or destructor cannot be declared 'thread'
1>D:\Users\nekit\Downloads\test\LightGBM\include\LightGBM/network.h(208): error C2483: 'LightGBM::Network::buffer_' : object with constructor or destructor cannot be declared 'thread'
1>D:\Users\nekit\Downloads\test\LightGBM\include\LightGBM/network.h(208): error C2483: 'private: static std::vector<char,std::allocator<char> > LightGBM::Network::buffer_' : object with constructor or destructor cannot be declared 'thread'
========== Build: 0 succeeded, 1 failed, 0 up-to-date, 0 skipped ==========
```

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -75,8 +75,8 @@ if(WIN32 AND MINGW)
 endif()
 
 if(MSVC)
-    if(MSVC_VERSION LESS 1800)
-        message(STATUS "The compiler ${CMAKE_CXX_COMPILER} has no C++11 support. Please use a newer msvc.")
+    if(MSVC_VERSION LESS 1900)
+        message(STATUS "The compiler ${CMAKE_CXX_COMPILER} has no C++11 support. Please use a newer MSVC.")
     endif()
 
     SET(variables

--- a/R-package/README.md
+++ b/R-package/README.md
@@ -92,7 +92,6 @@ You will need:
 
 In addition, if you are using a Visual Studio precompiled DLL, assuming you do not have Visual Studio installed (if you have it installed, ignore the warnings below):
 
-* Visual Studio 2013 precompiled DLL: download and install Visual Studio Runtime for [2013](https://support.microsoft.com/en-us/help/3179560/update-for-visual-c-2013-and-visual-c-redistributable-package) (you will get an error about MSVCP120.dll missing otherwise)
 * Visual Studio 2015/2017 precompiled DLL: download and install Visual Studio Runtime for [2015](https://www.microsoft.com/en-us/download/details.aspx?id=52685)/[2017](https://go.microsoft.com/fwlink/?LinkId=746572) (you will get an error about MSVCP140.dll missing otherwise)
 
 Once you have all this setup, you can use `lgb.dl` from `lgbdl` package to install LightGBM from repository.

--- a/R-package/src/install.libs.R
+++ b/R-package/src/install.libs.R
@@ -57,7 +57,7 @@ if (!use_precompile) {
     } else {
       try_vs <- 0
       local_vs_def <- ""
-      vs_versions <- c("Visual Studio 15 2017 Win64", "Visual Studio 14 2015 Win64", "Visual Studio 12 2013 Win64")
+      vs_versions <- c("Visual Studio 15 2017 Win64", "Visual Studio 14 2015 Win64")
       for(vs in vs_versions){
         vs_def <- paste0(" -G \"", vs, "\"")
         tmp_cmake_cmd <- paste0(cmake_cmd, vs_def)

--- a/docs/Installation-Guide.rst
+++ b/docs/Installation-Guide.rst
@@ -16,7 +16,7 @@ Visual Studio (or MSBuild)
 With GUI
 ********
 
-1. Install `Visual Studio`_.
+1. Install `Visual Studio`_ (2015 or newer).
 
 2. Download `zip archive`_ and unzip it.
 
@@ -31,7 +31,7 @@ The exe file will be in ``LightGBM-master/windows/x64/Release`` folder.
 From Command Line
 *****************
 
-1. Install `Git for Windows`_, `CMake`_ (3.8 or higher) and `MSBuild`_ (**MSBuild** is not needed if **Visual Studio** is installed).
+1. Install `Git for Windows`_, `CMake`_ (3.8 or higher) and `MSBuild`_ (**MSBuild** is not needed if **Visual Studio** (2015 or newer) is installed).
 
 2. Run the following commands:
 
@@ -129,7 +129,7 @@ With GUI
 
 1. You need to install `MS MPI`_ first. Both ``msmpisdk.msi`` and ``MSMpiSetup.exe`` are needed.
 
-2. Install `Visual Studio`_.
+2. Install `Visual Studio`_ (2015 or newer).
 
 3. Download `zip archive`_ and unzip it.
 
@@ -146,7 +146,7 @@ From Command Line
 
 1. You need to install `MS MPI`_ first. Both ``msmpisdk.msi`` and ``MSMpiSetup.exe`` are needed.
 
-2. Install `Git for Windows`_, `CMake`_ (3.8 or higher) and `MSBuild`_ (MSBuild is not needed if **Visual Studio** is installed).
+2. Install `Git for Windows`_, `CMake`_ (3.8 or higher) and `MSBuild`_ (MSBuild is not needed if **Visual Studio** (2015 or newer) is installed).
 
 3. Run the following commands:
 
@@ -236,7 +236,7 @@ If you use **MinGW**, the build procedure are similar to the build in Linux. Ref
 
 Following procedure is for the MSVC (Microsoft Visual C++) build.
 
-1. Install `Git for Windows`_, `CMake`_ (3.8 or higher) and `MSBuild`_ (MSBuild is not needed if **Visual Studio** is installed).
+1. Install `Git for Windows`_, `CMake`_ (3.8 or higher) and `MSBuild`_ (MSBuild is not needed if **Visual Studio** (2015 or newer) is installed).
 
 2. Install **OpenCL** for Windows. The installation depends on the brand (NVIDIA, AMD, Intel) of your GPU card.
 
@@ -250,8 +250,6 @@ Following procedure is for the MSVC (Microsoft Visual C++) build.
 
    **Note**: match your Visual C++ version:
    
-   Visual Studio 2013 -> ``msvc-12.0-64.exe``,
-
    Visual Studio 2015 -> ``msvc-14.0-64.exe``,
 
    Visual Studio 2017 -> ``msvc-14.1-64.exe``.

--- a/python-package/README.rst
+++ b/python-package/README.rst
@@ -17,7 +17,7 @@ Note: 32-bit Python is not supported. Please install 64-bit version.
 
 Install from `PyPI <https://pypi.python.org/pypi/lightgbm>`_ Using ``pip``
 ''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''
-For Windows users, `VC runtime <https://go.microsoft.com/fwlink/?LinkId=746572>`_ is needed if Visual Studio (2013, 2015 or 2017) is not installed.
+For Windows users, `VC runtime <https://go.microsoft.com/fwlink/?LinkId=746572>`_ is needed if Visual Studio (2015 or 2017) is not installed.
 
 Install `wheel <http://pythonwheels.com>`_ via ``pip install wheel`` first. After that download the wheel file and install from it:
 

--- a/python-package/setup.py
+++ b/python-package/setup.py
@@ -98,7 +98,7 @@ def compile_cpp(use_mingw=False, use_gpu=False):
             lib_path = "../compile/windows/x64/DLL/lib_lightgbm.dll"
             if not use_gpu:
                 logger.info("Starting to compile with MSBuild from existing solution file.")
-                platform_toolsets = ("v141", "v140", "v120")
+                platform_toolsets = ("v141", "v140")
                 for pt in platform_toolsets:
                     status = silent_call(["MSBuild", "../compile/windows/LightGBM.sln",
                                           "/p:Configuration=DLL",
@@ -111,7 +111,7 @@ def compile_cpp(use_mingw=False, use_gpu=False):
                 if status != 0 or not os.path.exists(lib_path):
                     logger.warning("Compilation with MSBuild from existing solution file failed.")
             if status != 0 or not os.path.exists(lib_path):
-                vs_versions = ("Visual Studio 15 2017 Win64", "Visual Studio 14 2015 Win64", "Visual Studio 12 2013 Win64")
+                vs_versions = ("Visual Studio 15 2017 Win64", "Visual Studio 14 2015 Win64")
                 for vs in vs_versions:
                     logger.info("Starting to compile with %s." % vs)
                     status = silent_call(cmake_cmd + ["-G", vs])

